### ISSUE TITLE
Onboard and Propose the tasks to initiate the First Sprint for the LLM interface to Open Virome.

### ISSUE THREAD

**Issue by Stufedpanda** [serratus-bio/open-virome#126]
### Overview

Biological graphs are vast and complex, with many different relationships and features under consideration. I seek to optimize graph traversal and add an intuitive user interface that can quickly extract meaningful data from a biological graph with LLMs. 

### Background / Context

This project will start with a narrower scope focused on eimeria and its local relationships on the graph database.

### Hypothesis

Can an LLM be implemented to assist with providing improved interpretability and accessibility of a biological graph?

### Experiment

1. Fine-tune a pre-trained model with graph data.
2. Prompt LLM to retrieve nodes that relate to or answer the query.
3. Verify the accuracy of retrieved nodes to known biological data.

### Controls

**Positive Controls**
- Retrieving all related SRAs to Eimeria
- Retrieving all related Biology Projects to Eimeria

**Negative Controls**
- Retrieving no relationships between Eimeria and an unrelated node

### Expected Outcome

A positive outcome for this experiment would be for the model to retrieve an amount of nodes above a tolerable threshold (TBD). Additionally, the ability to summarize and answer queries accurately from the user prompt would be an accomplishment.

### Open Questions

1. How does the model function when given broad queries (i.e prompts that require a scope including the entirety of the graph) and if needed, how can we improve that function?

2. How can the model be improved? (training, preprocessing, different architectures)

**Comment by lukepereira** [comment]
## Task 1:

Let me know if you disagree, but I don't think we need to worry about the retrieval step and can focus more on the summarization and interpretation part. The current app already has a good interface for directly querying into the data, plus going from language -> similar entity search will need a lot more set up (i.e. a vector db for local search, community summaries for global search)

For now, a good start would be to set up a notebook that does the following:
1. Hard code the initial "Eimeria" SRA runs as an input
   ```
    MATCH (n:SRA)-[:HAS_HOST_METADATA]-(t:Taxon)
    WHERE t.taxGenus = 'Eimeria'
    RETURN n
   ```
2. Write graph queries to fetch all associated sOTUs, bioprojects, tissues, STAT hosts, biosample metadata etc. directly related to those SRA runs. Potentially include other indirectly related nodes.
3. Format all the above context data into a prompt that can used by the LLM. I've read that YAML format is popular for LLM prompting, but you can read more into it.
4. Add the final prompt, which should involve interpreting the context to answer a specific question
5. Evaluate how well the LLM's answer matches the expected response

Examples of final prompt:
- "Summarize the main areas of research associated to Eimeria BioProjects"
- "Summarize the tissue tropism of Eimeria"
- Read through some bioproject descriptions and cherry pick a specific answer you expect, i.e. "what are the molecular mechanisms governing the drug resistance of Eimeria" (from PRJNA471055)
- "What other hosts have the same viruses as Eimeria?"


If these perform well, we can potentially plug this into the app either as wikipedia-like summaries or allow free-text queries about a specific virome. I think it would be especially useful for parsing through bioprojects with specific questions. After we get some understanding of the capabilities we can circle back to the retrieval search step.

**Comment by lukepereira** [comment]
We'll also want to to test the behaviour of Kadipiro virus related runs/bioprojects, since this is likely going to the be use case we explore in the paper.
```
MATCH (s:SOTU)--(n:SRA)
WHERE s.taxSpecies =~ '.*Kadipiro.*'
RETURN n
```

**Comment by almosnow** [comment]
* Clustering around genera
* BioProject abstracts might not be informative

- [ ] Issue 1
- [ ] Issue 2
- [ ] Issue 3

**Comment by ababaian** [comment]
### Proposal (From Forbes)

Cataloging the life on our planet is an arduous task, especially when looking at viromes.
Nevertheless, many have taken that challenge head on to create systems of classification and
metadata. In light of recent technological advancements, such data is now stored virtually within ever expanding databases. As the amount of data increases, more results will turn up for queries, rendering the task of looking through them all an inanity for any reasonable researcher. My goal for this research is to develop a large language model (LLM), in particular a GraphRAG [1][2] model, to parse through one such database used in the RNAlab for user querying. This will allow the user to grab data about their topic or answers to questions through plain sentences rather than convoluted queries to get what they need. The reasoning behind the RAG architecture is to allow for the model to be dynamic with the addition of new data in the database, rather than needing to be retrained every time a new batch of data is entered. GraphRAG is designed to be used alongside graphs, which neo4j databases utilize, hence our model architecture is chosen. The model will consist of two parts, one for local search and one for global search.

Our first hypothesis questions the validity of using an AI model for querying a database
at all, and whether an LLM can effectively retrieve and interpret sample metadata and research
abstracts from a given local cluster of nodes in the graph. I plan to use the genus Eimeria and the Kadipiro species for testing the local search capabilities, giving the LLM all the nodes related to Eimera or Kadipiro, a prompt, and judging the quality of the output. Here I’d use multiple metrics for general LLM evaluation as well as methods of evaluating contextual answers specific to the data the model is given. In terms of general LLM performance, metrics like precision, recall, f1 score and ROUGE-n [5] can be used to measure output. For context-related LLM performance, tools like cosine-similarity and BERTscore can measure contextual similarities of embeddings. If the evaluation scores of the LLM outputs are deemed sufficient, the next step is for the model to be able to retrieve its own information from the database, in which requires further evaluation with precision and recall as well as the ranking of the importance of specific pieces of data. There exist many ways of ranking, such as TF-IDF and more recently, BM-25 [7], and many ways to evaluate these rankings, with tools such as precision@k, Mean average precision (MAP), DCG@k [8] or TREC-eval [9].

For global search, the database will be clustered into unlabeled communities. The LLM
will generate summaries for each community. When queried, the model uses the summaries to
produce an intermediate response [1][2] that are summary points of interest per community,
which are scored and sorted by rank. The most relevant points (highest scoring) are then used to generate several aggregate responses, which are then used as context for our model to generate a final output. Here we want to test the performance of our model in retrieving and contextualizing relevant data from the communities to answer the queries. Metrics used to evaluate our local search retrieval relevance can be reused here, such as MAP, nDCG and cosine similarity but instead put into a community or cluster context rather than individual nodes from our graph.

If successful, this model can be integrated into the Open Virome website to allow users to
query the database as a whole [Figures B1, B3], or to supplement their own queries of the
database [Figure B2]. If proven effective over time, many such models may be adapted to other
databases of similar or larger scale, to help reduce the overhead in information retrieval.

**Comment by ababaian** [comment]
> This will allow the user to grab data about their topic or answers to questions through plain sentences rather than convoluted queries to get what they need.

## Task 2: Define 5 exact questions/queries you would like to solve.
I'd like you to develop 5 "Example User Queries" which solve ask a real biological question, and each one should have a dependency set of different meta-data. In this way you can provide concrete examples of the quality and type of data which is being accessed, and how that addresses the problem. Open a `documentation` issue on the Open Virome repo for this.

For Example: "Show me all datasets in which an arbovirus is found in brain tissue."

The key points here are `Show me all datasets` which is requesting a list of `SRA runs`, does this implicitly entail a summary of their aggregate properties, such as number, technology type, and the meta-data which is available in the `SRA Run Info` tables?

`in which an arbovirus is found` is requesting for a specific class of viruses, namely mosquito-infecting viruses which are Arboviruses. We currently don't have a list of "Arboviruses" anywhere, but this could be generated from the data by intersecting all viruses identified in any dataset belonging to Mosquito taxa. This would be joining `SRA Run Info` table `organism` label, with NCBI taxonomy tables, under the family `Culicidae` (mosquito), then using that list of viruses (Serratus/Logan tables), that would be the "Arbovirus" definition.

`found in brain tissue` this would be a list of SRA runs which have associated tissue metadata matching brain, or a brain-related term such as Neuron, Cerebellum, all of which are stored in the `BRENDA` ontology tables/graph.

At  the heart, this question is asking for an intersection of viruses found in at least one dataset where the label is a child member of the `Culcidae` taxonomic family. That list is then intersected with the list of all brain- or brain-associated tissue terms, and those datasets would be reported as the set of "potential Arboviruses".

This is a critically important question because it allows one to ask  the related question, "Is this unknown virus which has showed up in this child at Sick Kids, an arbovirus which be linked to the patient's present fever and encephalitis?"

**Comment by ababaian** [comment]
## Task 3: Define a falsifiable, measurable hypothesis.
>Our first hypothesis questions the validity of using an AI model for querying a database
at all, and whether an LLM can effectively retrieve and interpret sample metadata and research
abstracts from a given local cluster of nodes in the graph.

There is no unit of measurement provided here, as such, there is no criteria by which the hypothesis can be measured to fail as an outcome of the proposed work. Leading into the hypothesis you have to establish what are the evaluation metric.

>In terms of general LLM performance, metrics like precision, recall, f1 score and ROUGE-n [5] can be used to measure output.

What is the strengths and weaknesses of each of these methods, how will you establish this measurement?

> tools like cosine-similarity and BERTscore can measure contextual similarities of embeddings.

Which embeddings?  These are not introduced yet.

> If the evaluation scores of the LLM outputs are deemed sufficient...
You must define the threshold of "sufficient" as an exact number, and this value must be defined prior to starting any experiment.

Sub-task is evident here. Create a Table with each evalutation criteria you can possible use for your task. You list many, for each one what are the:
- Required input to be provided
- Assumptions of the model/evaluation metric
- Tools which implement this evaluation metric
- Threshold values which are accepted as "success" criteria and "fail" criteria and "uncertain" criteria as they are used in the literature (find examples).
- Limitations or drawbacks of this method

**Comment by ababaian** [comment]
I think with Luke and what I've outlined, you have your three tasks for the sprint, this is exactly on the right track to on-board you. For each of our three suggestions Create an issue (you can directly copy-paste the text from here for these issues), define what your outcomes are, and prioritize them as you work on them.

To close this issue; you need to create `3` issues for each defined task above. Luke or I should :+1:  on the issue that it's documented to be on the `Ready` category of the project board (https://github.com/orgs/serratus-bio/projects/5/views/1). You can then choose whichever you think is the highest logical priority, set that to `DEFCON 2` and move it to the `in progress` category, the two other ones set them to `DEFCON 3/4` as you see fit.

**Comment by ababaian** [comment]
Great use of sketching ![Image](https://github.com/user-attachments/assets/fe36757e-2110-4bb4-a129-4b12c7b3379c)
